src/writer.rs

```diff
@@ -17,196 +17,259 @@ use crate::{
     DataElement, NiftiHeader, NiftiType, Result,
 };
 
-/// Write a nifti file (.nii or .nii.gz).
-///
-/// If a `reference` is given, it will be used to fill most of the header's fields. Otherwise, a
-/// default `NiftiHeader` will be built and written. In all cases, the `dim`, `datatype` and
-/// `bitpix` fields will depend only on `data`, not on the header. This means that the `datatype`
-/// defined in `reference` will be ignored. Because of this, `scl_slope` will be set to 1.0 and
-/// `scl_inter` to 0.0.
-pub fn write_nifti<P, A, S, D>(
-    header_path: P,
-    data: &ArrayBase<S, D>,
-    reference: Option<&NiftiHeader>,
-) -> Result<()>
-where
-    P: AsRef<Path>,
-    S: Data<Elem = A>,
-    A: DataElement,
-    A: TriviallyTransmutable,
-    D: Dimension + RemoveAxis,
-{
-    let compression_level = Compression::fast();
-    let is_gz = is_gz_file(&header_path);
-    let (header, data_path) =
-        prepare_header_and_paths(&header_path, data, reference, A::DATA_TYPE)?;
-
-    // Need the transpose for fortran ordering used in nifti file format.
-    let data = data.t();
-
-    let header_file = File::create(header_path)?;
-    if header.vox_offset > 0.0 {
-        if is_gz {
-            let mut writer = ByteOrdered::runtime(
-                GzEncoder::new(header_file, compression_level),
-                header.endianness,
-            );
-            write_header(writer.as_mut(), &header)?;
-            write_data::<_, A, _, _, _, _>(writer.as_mut(), data)?;
-            let _ = writer.into_inner().finish()?;
-        } else {
-            let mut writer = ByteOrdered::runtime(BufWriter::new(header_file), header.endianness);
-            write_header(writer.as_mut(), &header)?;
-            write_data::<_, A, _, _, _, _>(writer, data)?;
+/// Options and flags which can be used to configure how a NIfTI image is written.
+#[derive(Debug, Clone, PartialEq)]
+pub struct WriterOptions<'a> {
+    /// Where to write the output image (header and/or data).
+    header_path: PathBuf,
+    /// If given, it will be used to fill most of the header's fields. Otherwise, a default
+    /// `NiftiHeader` will be built and written. In all cases, the `dim`, `datatype` and `bitpix`
+    /// fields will depend only on `data` parameter passed to the `write` function, not on this
+    /// field. This means that the `datatype` defined in `reference` will be ignored. Because of
+    /// this, `scl_slope` will be set to 1.0 and `scl_inter` to 0.0.
+    reference: Option<&'a NiftiHeader>,
+    /// Whether to write the header and data in distinct files. (nii vs hdr+img)
+    header_file: bool,
```

Comment:
So far the nomenclature has narrowed to either "NIfTI file pair" or "header and volume files". Both should be clear, given the context.

On the other hand, the name `header_file` can be a bit confusing, considering that we also have `header_path`. A better name IMO is `write_header_file`.